Userland/Libraries/LibWeb/Crypto/SubtleCrypto.cpp

```diff
@@ -184,6 +184,100 @@ JS::NonnullGCPtr<JS::Promise> SubtleCrypto::digest(AlgorithmIdentifier const& al
+        if (key_data.has<Bindings::JsonWebKey>()) {
+            return realm.vm().throw_completion<JS::TypeError>(JS::ErrorType::NotAnObjectOfType, "BufferSource");
+        }
+
+        // 2. Let keyData be the result of getting a copy of the bytes held by the keyData parameter passed to the importKey() method.
+        auto data_buffer_or_error = WebIDL::get_buffer_source_copy(*key_data.get<JS::Handle<WebIDL::BufferSource>>()->raw_object());
+        if (data_buffer_or_error.is_error()) {
+            auto error = WebIDL::OperationError::create(realm, "Failed to copy bytes from ArrayBuffer"_fly_string);
+            auto promise = JS::Promise::create(realm);
+            promise->reject(error.ptr());
+            return promise;
+        }
+        real_key_data = data_buffer_or_error.value();
+    }
+
+    if (format == Bindings::KeyFormat::Jwk) {
+        // 1. If the keyData parameter passed to the importKey() method is not a JsonWebKey dictionary, throw a TypeError.
+        if (!key_data.has<Bindings::JsonWebKey>()) {
+            return realm.vm().throw_completion<JS::TypeError>(JS::ErrorType::NotAnObjectOfType, "JsonWebKey");
+        }
+
+        // 2. Let keyData be the keyData parameter passed to the importKey() method.
+        real_key_data = key_data.get<Bindings::JsonWebKey>();
+    }
+
+    // NOTE: The spec jumps to 5 here for some reason?
+    // 5. Let normalizedAlgorithm be the result of normalizing an algorithm, with alg set to algorithm and op set to "importKey".
+    auto normalized_algorithm = normalize_an_algorithm(algorithm, "importKey"_string);
+
+    // 6. If an error occurred, return a Promise rejected with normalizedAlgorithm.
+    if (normalized_algorithm.is_error()) {
+        auto promise = JS::Promise::create(realm);
+        auto error = normalized_algorithm.release_error();
+        auto error_value = error.value().value();
+        promise->reject(error_value);
+        return promise;
+    }
+
+    // 7. Let promise be a new Promise.
+    auto promise = JS::Promise::create(realm);
+
+    // 8. Return promise and perform the remaining steps in parallel.
+    // FIXME: We don't have a good abstraction for this yet, so we do it in sync.
+
+    // 9. If the following steps or referenced procedures say to throw an error, reject promise with the returned error and then terminate the algorithm.
+
+    // 10. Let result be the CryptoKey object that results from performing the import key operation
+    // specified by normalizedAlgorithm using keyData, algorithm, format, extractable and usages.
+    if (normalized_algorithm.release_value().name != "PBKDF2"sv) {
+        auto error = WebIDL::NotSupportedError::create(realm, MUST(String::formatted("Invalid algorithm '{}'", normalized_algorithm.release_value().name)));
+        promise->reject(error.ptr());
+        return promise;
+    }
+    auto maybe_result = pbkdf2_import_key(real_key_data, algorithm, format, extractable, key_usages);
```

Comment:
I feel like we have a TRY_OR_REJECT macro or two laying around when working with promises like this.